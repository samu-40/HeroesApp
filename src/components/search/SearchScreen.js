import React, { useMemo } from 'react';
import HeroeCard from '../heroes/HeroeCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router-dom';
import { parse } from "query-string";
import { getHeroeByName } from '../selectors/getHeroeByName';

import './search_screen.css';

const SearchScreen = ({ history }) => {

    const location = useLocation();
    
    const {q = ''} = parse(location.search);


    const [formValue, hadleInputChange] = useForm({textSearch: q});
    
    const {textSearch} = formValue;


    const heroesFiltered = useMemo(() => getHeroeByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${textSearch}`);
    };

    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>

            <div className='container-search-screen'>
                <div className='container-search'>
                    <form 
                        onSubmit={handleSearch}
                        className='search'
                    >
                        <input
                            type='text'
                            placeholder='Find your heroe'
                            onChange={hadleInputChange}
                            name='textSearch'
                            value={textSearch}
                            autoComplete='off'
                        />

                        <button
                            type='submit'
                        >
                            Search
                        </button>
                    </form>
                </div>

                <div className='results'>
                    <h4>Results</h4>
                    <hr/>

                    {
                        (q === '') &&
                            <div>Search a heroe</div>
                    }

                    {
                        (q !== '' && heroesFiltered.length === 0) &&
                            <div id='error'>
                                {`Hero ${q} not found`}
                            </div>
                    }

                    {
                        heroesFiltered.map(hero => (
                            <HeroeCard 
                                key={hero.id}
                                {...hero}
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}

export default SearchScreen
