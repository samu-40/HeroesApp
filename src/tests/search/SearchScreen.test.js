import '../../setupTest';
import { mount } from "enzyme";
import SearchScreen from '../../components/search/SearchScreen';
import { MemoryRouter, Route } from 'react-router-dom';

describe('Pruebas en SearchScreen.js', () => {
        
    test('Mostrar componente con valores por defecto', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route 
                    path='/search' 
                    component={SearchScreen}
                />
            </MemoryRouter>
        );
        
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('h4').text().trim()).toBe('Results');

    });

    test('Debe mostrar al heroe y el input con el valor del queryString', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                    path='/search' 
                    component={SearchScreen}
                />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();

    });

    test('Deme mostrar un error si no se encuentra al heroe', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman1253']}>
                <Route 
                    path='/search' 
                    component={SearchScreen}
                />
            </MemoryRouter>
        );

        expect(wrapper.find('#error').text().trim()).toBe('Hero batman1253 not found');
        expect(wrapper).toMatchSnapshot();

    });

    test('Debe llamar al push del history', () => {
        
        const history = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                    path='/search' 
                    component={ () => <SearchScreen history={history}/> }
                />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'textSearch',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(history.push).toHaveBeenCalledWith('?q=batman');

    });    
    
});
