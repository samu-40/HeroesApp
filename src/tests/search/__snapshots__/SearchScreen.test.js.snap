// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScreen.js Debe mostrar al heroe y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="container-search-screen"
  >
    <div
      className="container-search"
    >
      <form
        className="search"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          name="textSearch"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="text"
          value="batman"
        />
        <button
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="results"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="main-container-flex"
      >
        <div
          className="img"
        >
          <img
            alt="Batman"
            className="picture"
            src="./assets/heroes/dc-batman.jpg"
          />
        </div>
        <div
          className="text"
        >
          <h3>
            Batman
          </h3>
          <p>
            Bruce Wayne
          </p>
          <a
            href="/hero/dc-batman"
            onClick={[Function]}
          >
            More...
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en SearchScreen.js Deme mostrar un error si no se encuentra al heroe 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="container-search-screen"
  >
    <div
      className="container-search"
    >
      <form
        className="search"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          name="textSearch"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="text"
          value="batman1253"
        />
        <button
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="results"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        id="error"
      >
        Hero batman1253 not found
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en SearchScreen.js Mostrar componente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="container-search-screen"
  >
    <div
      className="container-search"
    >
      <form
        className="search"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          name="textSearch"
          onChange={[Function]}
          placeholder="Find your heroe"
          type="text"
          value=""
        />
        <button
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="results"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div>
        Search a heroe
      </div>
    </div>
  </div>
</div>
`;
